language/tools/read-write-set/src/lib.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use move_core_types::{identifier::Identifier, language_storage::ModuleId};
+use move_model::model::GlobalEnv;
+use prover_bytecode::{
+    function_target_pipeline::{FunctionTargetPipeline, FunctionTargetsHolder, FunctionVariant},
+    read_write_set_analysis::{ReadWriteSetProcessor, ReadWriteSetState},
+};
+use vm::file_format::CompiledModule;
+
+pub struct ReadWriteSet {
+    targets: FunctionTargetsHolder,
+    env: GlobalEnv,
+}
+
+/// Infer read/write set results for `modules`.
+/// The `modules` list must be topologically sorted by the dependency relation
+/// (i.e., a child node in the dependency graph should appear earlier in the
+/// vector than its parents), and all dependencies of each module must be
+/// included.
+pub fn analyze<'a>(modules: impl IntoIterator<Item = &'a CompiledModule>) -> Result<ReadWriteSet> {
```

Comment:
I was wondering instead of taking in a module iterator as input, will it make better sense to take a storage reader (say given a ModuleId returns a CompiledModule) will make a better sense for production? As in the long term this api will be a bit more relaxed and we won't be able to pull the entire list of modulesin the long run. Another alterative is to put the burden onto the caller side: the caller will need to fetch all transitive dependencies of modules to be invoked within a block. This approach will be less pratical as we have more modules but should be fine in a foreseable future.

Reply:
In the current approach, we 
1. Use `move-model` to builds a `GlobalEnv` from a dependency-complete list of modules
2. Use the `GlobalEnv` to build a call graph (guaranteed to be a DAG in the absence of recursion) and begin compositional read/write set analysis bottom-up from the leaves to the roots, analyzing each callee before its callers.
3. Use the results of (2) to return the read/write set of a particular procedure (e.g. an invoked script function).


Steps (1) and (2) fundamentally require building an env for + analyzing the dependencies of the procedure whose output we are interested in; there's no way around that. It would be possible (though not easy) to redesign steps (1) and (2) to be a bit lazier; I have done such things in other static analysis projects at FB. But that point at which you need such techniques is usually when you have >10M LOC. We are a long way off from that in Diem, and we may never get there given the stringent perf requirements of transactions O(milliseconds). So I think an iterator of modules is just fine for years to come.